src/coreclr/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/ICastableHelpers.cs

```diff
@@ -22,5 +23,30 @@ internal static RuntimeType GetImplType(ICastable castable, RuntimeType interfac
         {
             return castable.GetImplType(new RuntimeTypeHandle(interfaceType)).GetRuntimeType();
         }
+
+        [Diagnostics.StackTraceHidden]
+        internal static RuntimeType? GetInterfaceImplementation(ICastableObject castableObject, RuntimeType interfaceType, bool throwIfNotFound)
+        {
+            RuntimeTypeHandle handle = castableObject.GetInterfaceImplementation(new RuntimeTypeHandle(interfaceType), throwIfNotFound);
+            if (handle.Equals(default))
+            {
+                if (throwIfNotFound)
+                    throw new InvalidCastException(SR.Format(SR.InvalidCast_FromTo, castableObject.GetType().ToString(), interfaceType.ToString()));
+
+                return null;
+            }
+
+            RuntimeType implType = handle.GetRuntimeType();
+            if (!implType.IsInterface)
+                throw new InvalidProgramException(SR.Format(SR.ICastableObject_NotInterface, implType.ToString()));
```

Comment:
The [documentation](https://docs.microsoft.com/en-us/dotnet/api/system.invalidprogramexception?view=netcore-3.1) says that InvalidProgramException means bad IL or metadata, and indicates bug in the compiler that produced the program. It is not the case here.

Would e.g. `InvalidOperationException` be more appropriate here?

Reply:
@elinor-fung I updated the proposal to use `InvalidOperationException`.